File: core/model/stock-domain-service.go
package model

import (
	"github.com/dc0d/wrapperr"
)

type EntityDomainService interface {
	UpdateStock(incoming PriceUpdated) error
}

type EntityDSBuilder struct {
	Repository Repository
	SeqGen     SeqGen
	TimeSource TimeSource
}

func (builder EntityDSBuilder) Build() *EntityDS {
	return &EntityDS{
		repo:    builder.Repository,
		seqGen:  builder.SeqGen,
		timeSrc: builder.TimeSource,
	}
}

type EntityDS struct {
	repo    Repository
	seqGen  SeqGen
	timeSrc TimeSource
}

func (ds *EntityDS) UpdateStock(incoming PriceUpdated) error {
	updated, err := ds.repo.PriceUpdatedEventExists(incoming.ID)
	if err != nil {
		return wrapperr.WithStack(err)
	}
	if updated {
		return ErrPriceAlreadyUpdated
	}

	oldStock, err := ds.repo.LoadStock(incoming.UpdatedStock.StockID)
	if err != nil {
		if err != ErrNotFound {
			return wrapperr.WithStack(err)
		}
		oldStock = Stock{
			ID:    incoming.UpdatedStock.StockID,
			Price: incoming.UpdatedStock.Price,
			Date:  ds.timeSrc.NowUTC(),
		}
	}

	var updatedStock Stock
	updatedStock.ID = incoming.UpdatedStock.StockID
	updatedStock.Rev = ds.seqGen.Next()
	updatedStock.Price = incoming.UpdatedStock.Price
	updatedStock.Date = ds.timeSrc.NowUTC()

	change := ChangeSet{
		Incoming:     incoming,
		UpdatedStock: updatedStock,
		Outgoing: StockUpdated{
			ID:        ds.seqGen.Next(),
			Timestamp: ds.seqGen.Next(),
			Stock:     updatedStock,
		},
		OldStockRev: oldStock.Rev,
	}

	return ds.repo.Submit(change)
}

const (
	ErrPriceAlreadyUpdated SentinelError = "stock price already updated"
)

File: stocks/src/core/model/stock-domain-service.ts
import { PriceUpdated } from "./events";
import { Repository, ChangeSet } from "./repository";
import { SeqGen } from "./seq-gen";
import { TimeSource } from "./time-source";
import { EventAlreadyHandledError, StockNotFoundError } from "./errors";
import { Stock } from "./stock";

export class StockDomainService {
  private readonly repo: Repository;
  private readonly idGen: SeqGen;
  private readonly timeSource: TimeSource;

  constructor(repo: Repository, idGen: SeqGen, timeSource: TimeSource) {
    this.repo = repo;
    this.idGen = idGen;
    this.timeSource = timeSource;
  }

  async updateStock(incoming: PriceUpdated): Promise<void> {
    if (await this.repo.priceUpdatedEventExists(incoming.id)) {
      throw new EventAlreadyHandledError();
    }

    let stock: Stock;
    try {
      stock = await this.repo.loadStock(incoming.updatedStock.stockId);
      stock = {
        ...stock,
        price: incoming.updatedStock.price,
      };
    } catch (error) {
      if (error instanceof StockNotFoundError) {
        stock = {
          id: incoming.updatedStock.stockId,
          rev: this.idGen.next(),
          price: incoming.updatedStock.price,
          date: this.timeSource.utcNow(),
        };
      } else {
        throw {
          error: error,
          message: "loadStock",
        };
      }
    }

    const changeSet = {
      incoming: incoming,
      updatedStock: stock,
      outgoing: {
        id: this.idGen.next(),
        timestamp: this.idGen.next(),
        stock: stock,
      },
    };

    await this.repo.submit(changeSet as ChangeSet);

    return;
  }
}
